src/libappimage/core/impl/TraversalType2.cpp

```diff
@@ -0,0 +1,333 @@
+    }
+
+    // Fetch first entry
+    next();
+}
+
+TraversalType2::~TraversalType2() {
+    sqfs_traverse_close(&trv);
+
+    clog << "Closing " << path << " as Type 2 AppImage" << endl;
+    sqfs_destroy(&fs);
+}
+
+void TraversalType2::next() {
+    sqfs_err err;
+    if (!sqfs_traverse_next(&trv, &err))
+        completed = true;
+
+    if (err != SQFS_OK)
+        throw IOError("sqfs_traverse_next error");
+
+    if (!completed) {
+        currentEntryType = readEntryType();
+        currentEntryPath = readEntryName();
+        currentEntryLink = currentEntryType == PayloadEntryType::LINK ? readEntryLink() : std::string();
+    } else {
+        currentEntryType = PayloadEntryType::UNKNOWN;
+        currentEntryPath = std::string();
+        currentEntryLink = std::string();
+    }
+}
+
+bool TraversalType2::isCompleted() const {
+    return completed;
+}
+
+std::string TraversalType2::getEntryName() const {
+    return currentEntryPath;
+}
+
+appimage::core::PayloadEntryType TraversalType2::getEntryType() const {
+    return currentEntryType;
+}
+
+void TraversalType2::extract(const std::string& target) {
+    sqfs_inode inode;
+    if (sqfs_inode_get(&fs, &inode, trv.entry.inode))
+        throw IOError("sqfs_inode_get error");
+
+    // create target parent dir
+    auto parentPath = bf::path(target).parent_path();
+    bf::create_directories(parentPath);
+
+    // handle each inode type properly
+    switch (inode.base.inode_type) {
+        case SQUASHFS_DIR_TYPE:
+        case SQUASHFS_LDIR_TYPE:
+            extractDir(target);
+            break;
+        case SQUASHFS_REG_TYPE:
+        case SQUASHFS_LREG_TYPE:
+            extractFile(inode, target);
+            break;
+        case SQUASHFS_SYMLINK_TYPE:
+        case SQUASHFS_LSYMLINK_TYPE:
+            extractSymlink(inode, target);
+            break;
+        default:
+            throw AppImageError("AppImage Type 2 inode.base.inode_type " +
+                                std::to_string(inode.base.inode_type) + " not supported yet");
+    }
+}
+
+sqfs_err TraversalType2::sqfsStat(sqfs* fs, sqfs_inode* inode, struct stat* st) {
+    // code borrowed from the "extract.c" file at squashfuse
+    sqfs_err err;
+    uid_t id;
+
+    memset(st, 0, sizeof(*st));
+    // fill stats
+    st->st_mode = inode->base.mode;
+    st->st_nlink = inode->nlink;
+    st->st_mtime = st->st_ctime = st->st_atime = inode->base.mtime;
+
+    if (S_ISREG(st->st_mode)) {
+        /* FIXME: do symlinks, dirs, etc have a size? */
+        st->st_size = inode->xtra.reg.file_size;
+        st->st_blocks = st->st_size / 512;
+    } else if (S_ISBLK(st->st_mode) || S_ISCHR(st->st_mode)) {
+        st->st_rdev = sqfs_makedev(inode->xtra.dev.major,
+                                   inode->xtra.dev.minor);
+    } else if (S_ISLNK(st->st_mode)) {
+        st->st_size = inode->xtra.symlink_size;
+    }
+
+    st->st_blksize = fs->sb.block_size; /* seriously? */
```

Comment:
What's the issue (aside from the comment being on the same line)? Is the line okay or not? If you are not sure, you should (rather must) add more context information, like e.g., why are you suspicious of this expression, or who might know that could be asked. This comment doesn't help me at all, it rather confuses and scares me.

Reply:
This is part of the code that I copied from the `squashfs`. Also, the comment came from there. I don't know either what they mean.

Reply:
Argh. Copied means they own the copyright on that code, right? That's mentioned nowhere! Please rewrite all copyrighted material from scratch (you may take inspiration from other code bases).

Reply:
Fixed